### PR TITLE
juju: Fix upgrade actions not working with resources

### PR DESCRIPTION
**Pull request by Cynerva**
**What this PR does / why we need it**:

This fixes an issue with the kubernetes-master and kubernetes-worker charms, where running the `upgrade` action does not actually perform an upgrade when snaps are attached as resources.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/528

**Special notes for your reviewer**:

The underlying issue is that both layer-snap and the kubernetes layers are using `any_file_changed` to look for changes in the resources. This PR fixes it by removing the use of `any_file_changed` in the top-level layers, and implementing our own code for it instead.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 1]
```diff
@@ -107,45 +109,75 @@ def check_for_upgrade_needed():
     add_rbac_roles()
     set_state('reconfigure.authentication.setup')
     remove_state('authentication.setup')
-    changed = snap_resources_changed()
-    if changed == 'yes':
-        set_upgrade_needed()
-    elif changed == 'unknown':
+
+    if not db.get('snap.resources.fingerprint.initialised'):
```
Comment (hyperbolic2346) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: nit: initialized

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 2]
```diff
@@ -88,12 +91,56 @@ def upgrade_charm():
     set_state('kubernetes-worker.restart-needed')
 
 
+def get_resource_checksum_db_key(resource):
+    ''' Convert a resource name to a resource checksum database key. '''
+    return 'kubernetes-worker.resource-checksums.' + resource
+
+
+def calculate_resource_checksum(resource):
```
Comment (hyperbolic2346) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: would be nice to have a common lib for these shared functions

**Review by ktsakalozos — APPROVED** [review]

**Pull request merged**